### ISSUE TITLE
Openssl-3.0.8 Degradation in speed test for few algorithms (noticed similar behaviour in 3.1.0-beta1 as well)

### ISSUE THREAD

**Issue by Ishani-3** [openssl/openssl#20391]
I have Upgraded Openssl from 1.1.1q to 3.0.8 version, enabled legacy providers by making changes in openssl.cnf file and then built. I'm able to see legacy and default providers by running list providers command. For few algorithms I notice degradation in speed test results and observed similar behaviour in 3.1.0-beta1 as well. 

```
Command: openssl list -providers
```
```
Providers:
  default
    name: OpenSSL Default Provider
    version: 3.0.8
    status: active
  legacy
    name: OpenSSL Legacy Provider
    version: 3.0.8
    status: active
```

Below are the algorithms for which degradation in speedtest is observed:

1. aes-192-ecb ( 64 and 256 bytes)
2. aes-256-ecb ( 64 bytes )
3. aes-128-ecb ( 256 bytes)
4. sha256 (16  and 64 bytes ) 
5. sha1 ( 16 , 64 and 256 bytes)
6. md5 (16 and 64 bytes)

Here are the speed test results for few of them :
```
```
command: openssl speed -evp $alg
```
1. version: 1.1.1q
Doing aes-256-ecb for 3s on 16 size blocks: 161439348 aes-256-ecb's in 2.99s
Doing aes-256-ecb for 3s on 64 size blocks: 171267332 aes-256-ecb's in 3.00s
Doing aes-256-ecb for 3s on 256 size blocks: 72685445 aes-256-ecb's in 3.00s
Doing aes-256-ecb for 3s on 1024 size blocks: 20627449 aes-256-ecb's in 3.00s
Doing aes-256-ecb for 3s on 8192 size blocks: 2666312 aes-256-ecb's in 3.00s
Doing aes-256-ecb for 3s on 16384 size blocks: 1345085 aes-256-ecb's in 3.00s
OpenSSL 1.1.1q  5 Jul 2022
built on: Wed Sep 14 18:24:04 2022 UTC
options:bn(64,64) rc4(16x,int) des(int) aes(partial) idea(int) blowfish(ptr) 
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAESNI_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DNDEBUG -DOPENSSL_NO_CAPIENG
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
aes-256-ecb     863889.49k  3653703.08k  6202491.31k  7040835.93k  7280809.30k  7345957.55k

version: 3.0.8
Doing AES-256-ECB for 3s on 16 size blocks: 187653837 AES-256-ECB's in 2.99s
Doing AES-256-ECB for 3s on 64 size blocks: 160868522 AES-256-ECB's in 3.00s
Doing AES-256-ECB for 3s on 256 size blocks: 69426665 AES-256-ECB's in 3.00s
Doing AES-256-ECB for 3s on 1024 size blocks: 20375664 AES-256-ECB's in 3.00s
Doing AES-256-ECB for 3s on 8192 size blocks: 2681469 AES-256-ECB's in 3.00s
Doing AES-256-ECB for 3s on 16384 size blocks: 1345833 AES-256-ECB's in 3.00s
version: 3.0.8
built on: Mon Feb 27 06:15:09 2023 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG -DOPENSSL_NO_CAPIENG
CPUINFO: OPENSSL_ia32cap=0xfffa32034f8bffff:0x27eb
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
AES-256-ECB    1004167.69k  3431861.80k  5924408.75k  6954893.31k  7322198.02k  7350042.62k
```
```
2.  version: 1.1.1q
Doing aes-192-ecb for 3s on 16 size blocks: 173826691 aes-192-ecb's in 2.99s
Doing aes-192-ecb for 3s on 64 size blocks: 184126354 aes-192-ecb's in 3.00s
Doing aes-192-ecb for 3s on 256 size blocks: 84422682 aes-192-ecb's in 3.00s
Doing aes-192-ecb for 3s on 1024 size blocks: 24228918 aes-192-ecb's in 3.00s
Doing aes-192-ecb for 3s on 8192 size blocks: 3130369 aes-192-ecb's in 3.00s
Doing aes-192-ecb for 3s on 16384 size blocks: 1570449 aes-192-ecb's in 3.00s
OpenSSL 1.1.1q  5 Jul 2022
built on: Wed Sep 14 18:24:04 2022 UTC
options:bn(64,64) rc4(16x,int) des(int) aes(partial) idea(int) blowfish(ptr) 
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAESNI_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DNDEBUG -DOPENSSL_NO_CAPIENG
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
aes-192-ecb     930176.27k  3928028.89k  7204068.86k  8270137.34k  8547994.28k  8576745.47k

version: 3.0.8
Doing AES-192-ECB for 3s on 16 size blocks: 196372951 AES-192-ECB's in 3.00s
Doing AES-192-ECB for 3s on 64 size blocks: 170686225 AES-192-ECB's in 2.99s
Doing AES-192-ECB for 3s on 256 size blocks: 78220030 AES-192-ECB's in 3.00s
Doing AES-192-ECB for 3s on 1024 size blocks: 23751805 AES-192-ECB's in 3.00s
Doing AES-192-ECB for 3s on 8192 size blocks: 3112425 AES-192-ECB's in 3.00s
Doing AES-192-ECB for 3s on 16384 size blocks: 1568315 AES-192-ECB's in 3.00s
version: 3.0.8
built on: Mon Feb 27 06:15:09 2023 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG -DOPENSSL_NO_CAPIENG
CPUINFO: OPENSSL_ia32cap=0xfffa32034f8bffff:0x27eb
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
AES-192-ECB    1047322.41k  3708404.68k  6674775.89k  8107282.77k  8498995.20k  8565090.99k
```
```
3. version: 1.1.1q
Doing sha256 for 3s on 16 size blocks: 9878456 sha256's in 3.00s
Doing sha256 for 3s on 64 size blocks: 6296399 sha256's in 2.99s
Doing sha256 for 3s on 256 size blocks: 3122766 sha256's in 3.00s
Doing sha256 for 3s on 1024 size blocks: 1027770 sha256's in 3.00s
Doing sha256 for 3s on 8192 size blocks: 142149 sha256's in 3.00s
Doing sha256 for 3s on 16384 size blocks: 71961 sha256's in 3.00s
OpenSSL 1.1.1q  5 Jul 2022
built on: Wed Sep 14 18:24:04 2022 UTC
options:bn(64,64) rc4(16x,int) des(int) aes(partial) idea(int) blowfish(ptr) 
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAESNI_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DNDEBUG -DOPENSSL_NO_CAPIENG
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
sha256           52685.10k   134772.42k   266476.03k   350812.16k   388161.54k   393003.01k

version: 3.0.8
Doing sha256 for 3s on 16 size blocks: 9188448 sha256's in 3.00s
Doing sha256 for 3s on 64 size blocks: 6049513 sha256's in 3.00s
Doing sha256 for 3s on 256 size blocks: 3087634 sha256's in 3.00s
Doing sha256 for 3s on 1024 size blocks: 1026107 sha256's in 2.99s
Doing sha256 for 3s on 8192 size blocks: 143391 sha256's in 3.00s
Doing sha256 for 3s on 16384 size blocks: 72247 sha256's in 3.00s
version: 3.0.8
built on: Mon Feb 27 06:15:09 2023 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG -DOPENSSL_NO_CAPIENG
CPUINFO: OPENSSL_ia32cap=0xfffa32034f8bffff:0x27eb
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
sha256           49005.06k   129056.28k   263478.10k   351415.91k   391553.02k   394564.95k
```

```
perl configdata.pm --dump
```

```
Command line (with current working directory = .):

    /XXYYZZ/perl ./Configure --prefix=/XXXXXX/install --openssldir=/XXXXXX/install enable-rc4 no-shared enable-ssl3 -DOPENSSL_NO_CAPIENG linux-x86_64

Perl information:

    /XXYYZZ/perl
    5.24.0 for x86_64-linux

Enabled features:

    aria
    asm
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    bulk
    cached-fetch
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    ec
    ec2m
    ecdh
    ecdsa
    engine
    err
    filenames
    gost
    idea
    legacy
    makedepend
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    ocb
    ocsp
    padlockeng
    pic
    pinshared
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    siphash
    siv
    sm2
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    ssl-trace
    static-engine
    stdio
    tests
    threads
    tls
    ts
    ui-console
    weak-ssl-ciphers
    whirlpool
    ssl3
    ssl3-method
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    acvp-tests          [cascade]        OPENSSL_NO_ACVP_TESTS
    afalgeng            [too-old-kernel] OPENSSL_NO_AFALGENG
    asan                [default]        OPENSSL_NO_ASAN
    buildtest-c++       [default]
    crypto-mdebug       [default]        OPENSSL_NO_CRYPTO_MDEBUG
    devcryptoeng        [default]        OPENSSL_NO_DEVCRYPTOENG
    dynamic-engine      [cascade]
    ec_nistp_64_gcc_128 [default]        OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                 [default]        OPENSSL_NO_EGD
    external-tests      [default]        OPENSSL_NO_EXTERNAL_TESTS
    fips                [default]
    fips-securitychecks [cascade]        OPENSSL_NO_FIPS_SECURITYCHECKS
    fuzz-afl            [default]        OPENSSL_NO_FUZZ_AFL
    fuzz-libfuzzer      [default]        OPENSSL_NO_FUZZ_LIBFUZZER
    ktls                [default]        OPENSSL_NO_KTLS
    loadereng           [cascade]        OPENSSL_NO_LOADERENG
    md2                 [default]        OPENSSL_NO_MD2 (skip crypto/md2)
    msan                [default]        OPENSSL_NO_MSAN
    rc5                 [default]        OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                [default]        OPENSSL_NO_SCTP
    shared              [option]
    trace               [default]        OPENSSL_NO_TRACE
    ubsan               [default]        OPENSSL_NO_UBSAN
    unit-test           [default]        OPENSSL_NO_UNIT_TEST
    uplink              [no uplink_arch] OPENSSL_NO_UPLINK
    zlib                [default]
    zlib-dynamic        [default]

Config target attributes:

    AR => "ar",
    ARFLAGS => "qc",
    CC => "gcc",
    CFLAGS => "-Wall -O3",
    CXX => "g++",
    CXXFLAGS => "-Wall -O3",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib",
    RC => "windres",
    asm_arch => "x86_64",
    bn_ops => "SIXTY_FOUR_BIT_LONG",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cflags => "-pthread -m64",
    cppflags => "",
    cxxflags => "-std=c++11 -pthread -m64",
    defines => [ "OPENSSL_BUILDING_OPENSSL" ],
    disable => [  ],
    dso_ldflags => "-Wl,-z,defs",
    dso_scheme => "dlfcn",
    enable => [ "afalgeng" ],
    ex_libs => "-ldl -pthread",
    includes => [  ],
    lflags => "",
    lib_cflags => "",
    lib_cppflags => "-DOPENSSL_USE_NODELETE -DL_ENDIAN",
    lib_defines => [  ],
    module_cflags => "-fPIC",
    module_cxxflags => undef,
    module_ldflags => "-Wl,-znodelete -shared -Wl,-Bsymbolic",
    multilib => "64",
    perl_platform => "Unix",
    perlasm_scheme => "elf",
    shared_cflag => "-fPIC",
    shared_defflag => "-Wl,--version-script=",
    shared_defines => [  ],
    shared_ldflag => "-Wl,-znodelete -shared -Wl,-Bsymbolic",
    shared_rcflag => "",
    shared_sonameflag => "-Wl,-soname=",
    shared_target => "linux-shared",
    thread_defines => [  ],
    thread_scheme => "pthreads",
    unistd => "<unistd.h>",

Recorded environment:

    AR =
    BUILDFILE =
    CC = gcc
    CFLAGS =
    CPPFLAGS =
    CROSS_COMPILE =
    CXX = g++
    CXXFLAGS = -w -O
    HASHBANGPERL =
    LDFLAGS =
    LDLIBS =
    OPENSSL_LOCAL_CONFIG_DIR =
    PERL = /XXYYZZ/perl
    RANLIB =
    RC =
    RCFLAGS =
    WINDRES =
    __CNF_CFLAGS =
    __CNF_CPPDEFINES =
    __CNF_CPPFLAGS =
    __CNF_CPPINCLUDES =
    __CNF_CXXFLAGS =
    __CNF_LDFLAGS =
    __CNF_LDLIBS =

Makevars:

    AR              = ar
    ARFLAGS         = qc
    CC              = gcc
    CFLAGS          = -Wall -O3
    CPPDEFINES      = OPENSSL_NO_CAPIENG
    CPPFLAGS        =
    CPPINCLUDES     =
    CXX             = g++
    CXXFLAGS        = -w -O
    HASHBANGPERL    = /XXYYZZ/perl
    LDFLAGS         =
    LDLIBS          =
    PERL            = /XXYYZZ/perl
    RANLIB          = ranlib
    RC              = windres
    RCFLAGS         =

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    Makefile

build file:

    Makefile

build file templates:

    Configurations/common0.tmpl
    Configurations/unix-Makefile.tmpl
```

Please suggest ways through which the speed test outputs can be improved in comparision to 1.1.1q.

**Comment by paulidale** [comment]
Did you run speed with the `-evp` option in 1.1.1?  If not, the comparison isn't equivalent.  3.0 always uses EVP.

3.0 also has additional overhead because of the provider architecture.  It generally will be slightly slower than 1.1.1, although there have been performance improvements in some of the assembly implementations.

**Comment by Ishani-3** [comment]
@paulidale  Thankyou for the response! we have used -evp option for both versions 1.1.1q and 3.0.8 (openssl speed -evp $alg).

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input
